WT-13886 Organise directory structure for modularity tools in WT

Siddhartha8899 opened the pull request:
Organise directory structure for modularity tools in WT

Siddhartha8899 commented:
As part of this PR I also wish to
- update the wt_defs.py
- update the requirements file such that we could get the package from wiredtiger/layercparse

ershov commented on dist/modularity/access_check/check_sources.py:
```diff
@@ -11,21 +11,29 @@
 import layercparse as lcp
-import wt_defs
+
+def load_wt_defs(rootPath):
```
I think, `load_wt_defs` should be a part of the library.

Siddhartha8899 replied:
Updated in [b59c9ce](https://github.com/wiredtiger/wiredtiger/pull/11335/commits/b59c9cec310c3af0787db6a4ec6596ee48b059f5)

ershov commented on dist/modularity/access_check/check_sources.py:
If we don’t anticipate adding more content to sub-directories soon, I’d prefer keeping a simpler directory structure. At this stage, it doesn’t seem like the handful of files in the `modularity` directory require additional organization.

Siddhartha8899 replied:
Updated in [b59c9ce](https://github.com/wiredtiger/wiredtiger/pull/11335/commits/b59c9cec310c3af0787db6a4ec6596ee48b059f5)

ershov reviewed (changes requested):
Thanks, Sid, for this PR!  
I really like this change and have a few in-line comments.  

Regarding pulling the parser library - do you think we could use a GitHub URL to always fetch the latest version?  
On the flip side, scripts using the library might depend on a specific version or behavior, and future changes could break older scripts.  

Another question: does the Python environment have a caching mechanism? Ideally, we wouldn’t want it to hit GitHub for every script invocation. If caching exists, how does it handle invalidation, or how often does it check for updates? Also, what’s the process for updating the library if it changes after the initial setup of the environment?

Siddhartha8899 commented on dist/modularity/mod_stat/scan_sources.py:
I tried git mv but it didnt work, however, somehow it is fixed now with the new commit.

Siddhartha8899 commented:
> Regarding pulling the parser library - do you think we could use a GitHub URL to always fetch the latest version?
On the flip side, scripts using the library might depend on a specific version or behavior, and future changes could break older scripts.

Another question: does the Python environment have a caching mechanism? Ideally, we wouldn’t want it to hit GitHub for every script invocation. If caching exists, how does it handle invalidation, or how often does it check for updates? Also, what’s the process for updating the library if it changes after the initial setup of the environment?

@ershov  I have made an attempt at this, ensuring all these concerns are taken care of ->
1) updated setup.py file in https://github.com/wiredtiger/layercparse/pull/1
2) Deleted requirements and updated init.sh

ershov commented on dist/access_check/init.sh:
I wonder why Git doesn't recognize these files as moved?

Siddhartha8899 replied:
Not quite sure.

ershov commented on dist/modularity/init.sh:
```diff
@@ -0,0 +1,18 @@
+REPO_URL="https://github.com/wiredtiger/layercparse.git"
+BRANCH="main"
+HASH_FILE=".layercparse_commit_hash"
+LATEST_HASH=$(git ls-remote "$REPO_URL" "$BRANCH" | awk '{print $1}')
```
Last thing here: time-limit this call

Siddhartha8899 replied:
updated [706a3ad](https://github.com/wiredtiger/wiredtiger/commit/706a3ad2b138912166d424a10d6fc4448b78601e)

ershov commented on dist/modularity/init.sh:
```diff
@@ -0,0 +1,28 @@
+CACHE_TIMEOUT=0.5
+LATEST_HASH_CMD="git ls-remote \"$REPO_URL\" \"$BRANCH\" | awk '{print \$1}'"
+if ($IS_CACHED); then
+    if ! LATEST_HASH=$(timeout "$CACHE_TIMEOUT"s bash -c "$LATEST_HASH_CMD"); then
```
`timeout` doesn't exist on MacOS. I suggest implementing this in bash.

Also, if we hit timeout, we still want to run a background update.

And can we use `stat` or something to compare timestamps so that we don't have to go to internet each time when we run it? There could be a minimal interval of 1 hour or 1 day or so.

```bash
#!/bin/bash

REPO_URL="https://github.com/wiredtiger/layercparse.git"
BRANCH="main"
TIMEOUT=0.5

is_cached() {
  pip3 -qq show layercparse
  # true # for debug
}

get_latest_hash() {
  #sleep 2 # for debug
  git ls-remote "$REPO_URL" "$BRANCH" | awk '{print $1}'
}

get_latest_hash_timed() {
  get_latest_hash &
  P1="$!"
  (
    sleep $TIMEOUT
    kill $P1 2>/dev/null
  ) &
  wait $P1
}

if is_cached; then
  echo "Cached"
  if HASH=`get_latest_hash_timed`; then
    echo "Got hash: $HASH"
  else
    if [[ $? -eq 143 ]]; then
      echo "Timeout. Getting hash in the background"
      (
        HASH=`get_latest_hash`
        echo "Background hash = $HASH"
        # ... update the environment
      ) 2>&1 > /dev/null &
    else
      echo "Failed to get hash"
    fi
  fi
else
  echo "Not cached"
  HASH=`get_latest_hash`
  echo "Hash: $HASH"
fi
```

Siddhartha8899 replied:
Updated in [371f6e7](https://github.com/wiredtiger/wiredtiger/pull/11335/commits/371f6e705827a8e4d9b4e169dc8a096a4e3fad65)

ershov reviewed (dismissed):
All looks great! Just one more change in the `init` script.

ershov commented on dist/modularity/init.sh:
```diff
@@ -0,0 +1,29 @@
+BRANCH="main"
+HASH_FILE=".venv/layercparse_hash"
+
+is_layercparse_cached() {
+  pip3 -qq show layercparse > /dev/null 2>&1
+}
+
+# Function to check if layercparse cache is older than 24 hours
+is_layercparse_cache_outdated() {
+    [[ -z $(find "$HASH_FILE" -mtime -1 2>/dev/null) ]]
+}
+
+if ! is_layercparse_cached || is_layercparse_cache_outdated; then
+    latest_hash=$(git ls-remote $REPO_URL $BRANCH | awk '{print $1}')
+    if [[ ! -f "$HASH_FILE" || "$latest_hash" != "$(cat $HASH_FILE)" || ! is_layercparse_cached ]]; then
+        pip3 -q --disable-pip-version-check install git+"$REPO_URL@$BRANCH"
+        echo "$latest_hash" > "$HASH_FILE"
+    fi
+    touch "$HASH_FILE"
+fi
```
```suggestion
is_layercparse_cached() {
    pip3 -qq show layercparse > /dev/null 2>&1
}

is_layercparse_cache_outdated() {
    # We can't use "stat" because the flags are different across platforms.
    # "find" will give no output if the file doesn't exist or is too old.
    [[ -z $(find "$HASH_FILE" -mtime -1 2>/dev/null) ]] && return 0

    # touch the file as we don't attempt to update more frequently than once a day.
    touch "$HASH_FILE"

    CACHED_HASH=$(cat "$HASH_FILE")
    REMOTE_HASH=$(git ls-remote "$REPO_URL" "$BRANCH" | awk '{print $1}')
    [[ -n "$REMOTE_HASH" && "$CACHED_HASH" != "$REMOTE_HASH" ]]  # Return the result of comparison
}

if ! is_layercparse_cached || is_layercparse_cache_outdated; then
    pip3 -q --disable-pip-version-check install git+"$REPO_URL@$BRANCH"
    echo "$latest_hash" > "$HASH_FILE"
fi
```

raviprakashgiri29 reviewed (approved):
As discussed with Sid. Changes LGTM
